### PR TITLE
Add native response file support

### LINKED ISSUE #846: Add native response file support
As a CLI developer, I want Swift Argument Parser to natively support response file so I can make use of them in my CLI.

https://bh0.phas.ubc.ca/C2/intel/docs/c_ug/linux209a.htm

### PR DESCRIPTION
**Pull request by bkhouri**
Add native response file support in Swift Argument Parser for all `ParsableCommand` and `AsyncParsableCommand` commands. When at least one `@file` reference is present in the input, parser error messages now include a multi-line `at <file>:<line>` / `included from <file>:<line>` block that pinpoints the offending argument and its full include chain.

Fixes: #846
Relates to #41 

### Checklist
- [x] I've added at least one test that validates that my change is working, if appropriate
- [x] I've followed the code style of the rest of the project
- [x] I've read the [Contribution Guidelines](https://github.com/apple/swift-argument-parser/blob/main/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary

### CODE REVIEW

File: Tests/ArgumentParserEndToEndTests/ResponseFileEndToEndTests.swift [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+        --name
+        @@notafile
+        """)
+
+    AssertParse(SimpleCommand.self, ["@\(responseFile)"]) { command in
+      XCTAssertEqual(command.name, "@notafile")
+    }
+  }
+
+  func testResponseFileWithTerminator() throws {
```
Comment (natecook1000) [inline Tests/ArgumentParserEndToEndTests/ResponseFileEndToEndTests.swift]: Let's have a test with mixed arguments, with a response file with a terminator first – does it still parse the inline arguments after the response file or treat them as post-terminator args?

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+--input input.txt
+--output output.txt
+--verbose --count 42
+```
+
+## Advanced Features
+
+### Quoted Arguments
+
+Response files support quoted arguments for values containing spaces:
```
Comment (cmcgee1024) [inline Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md]: question: quoting brings up the question of how escaping works for things like spaces in unquoted text, and escaping the quotes themselves. If you intend on using shell conventions, then what about single quotes, or escaping scapes with `\`?

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+```
+--input=input.txt
+--output=output.txt
+--count=42
+```
+
+### Nested Response Files
+
+Response files can reference other response files:
```
Comment (cmcgee1024) [inline Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:216]: suggestion: Maybe don't add nesting unless there's a compelling workflow to do so. By having response files it alleviates the problem of command lines that are too long for the OS, improves versioning, and clarity. I'm not sure if nesting adds much beyond that, and brings more levels of indirection for the user to figure out where an argument came from.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+Arguments are processed in order, so later arguments can override earlier ones:
+
+```bash
+# If base-config.txt contains "--count 10"
+mycommand @base-config.txt --count 20  # Final count will be 20
+```
+
+## Literal At Signs
+
+To use a literal `@` character in an argument, escape it with `@@`:
```
Comment (cmcgee1024) [inline Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:336]: praise: it is good to call this kind of escaping out like this.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+```bash
+mycommand --username @@admin  # Passes "@admin" as the username
+```
+
+In response files:
+
+```
+--message @@important: This is critical
```
Comment (cmcgee1024) [inline Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:345]: thought: if nesting isn't required then this kind of escaping becomes unnecessary in the response files themselves.

**Comment by rauhul** [comment]
IMO this sort of feature will need to come with some sort of source location feature too

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 3 above)
Comment (allevato) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:216]: LLVM's response file parsing supports nesting. I know that swift-driver _today_ uses its own parsing instead of swift-argument-parser, but if there's any desire in the future to move existing LLVM-based tools from C++ onto Swift using swift-argument-parser, we'd need nested response file support to keep parity.

IMO, we should support it unless there's a compelling reason *not* to, to avoid unnecessary limitations on what the library can do. It looks like it falls out naturally from the implementation.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 2 above)
Comment (thomasvl) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md]: https://github.com/apple/swift-argument-parser/issues/41 has some links to how this is supported by some other languages/packages - some of them let one control the support the files will support.

Since someone could be using this support to migrate a tool to Swift, they might need the quoted form while some other migration might need the shell escaping support.

Likewise for what the marker on the command line is to say a file is a params/response file. So it if the leading `@` maybe should be configurable.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 3 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:216]: The change currently supports nested response file.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 5 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:345]: we are supporting nested response files.  See [this conversation](https://github.com/apple/swift-argument-parser/pull/909#discussion_r3318935465).

**Comment by bkhouri** [comment]
> IMO this sort of feature will need to come with some sort of source location feature too

Thanks for this feedback.  A source location feature has been added.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 2 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md]: The documents has been updated to provide additional information on the quotes arguments.

**Comment by bkhouri** [comment]
marking as `Draft` to prevent accidentally merging before it's ready.

File: Sources/ArgumentParser/Documentation.docc/Articles/ExperimentalFeatures.md [hunk 6]
```diff
@@ -13,6 +13,7 @@ If you have any feedback on experimental features, please [open a GitHub issue][
 | Name | Description | related PRs | Version |
 | ------------- | ------------- | ------------- | ------------- |
 | `--experimental-dump-help`  | Dumps command/argument/help information as JSON | [#310][] [#335][] | 0.5.0 or newer |
+| `--experimental-dump-arguments-source-location`  | Dumps the parsed argument tree with each value's source location (file:line for response-file args, `argv[N]` for command-line args). Accepts `=text` (default) or `=json`. | — | 1.9.0 or newer |
```
Comment (bkhouri) [inline Sources/ArgumentParser/Documentation.docc/Articles/ExperimentalFeatures.md]: chore: need to update the version where this feature is available. 

@natecook1000: could you please confirm what version I should set this to? Is there a "Next" equivalent that is used in this project?

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+Arguments are processed in order, so later arguments can override earlier ones:
+
+```bash
+# If base-config.txt contains "--count 10"
+mycommand @base-config.txt --count 20  # Final count will be 20
+```
+
+## Literal At Signs
+
+To use a literal `@` character in an argument, escape it with `@@`:
```
Comment (bkhouri) [inline Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:336]: Chore: Add a reference back to the quoted strings as another way to add a literal `@`.

File: Sources/ArgumentParser/Documentation.docc/Articles/ExperimentalFeatures.md (hunk 6 above)
Comment (bkhouri) [inline Sources/ArgumentParser/Documentation.docc/Articles/ExperimentalFeatures.md]: chore (blocking): Add the related PR.

Add `[#909](https://github.com/apple/swift-argument-parser/pull/909)` as a related PR.

Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ExperimentalFeatures.md]: I updated the version to `unreleased`

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 7 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:336]: done.

File: Sources/ArgumentParser/Parsing/SplitArguments.swift [hunk 8]
```diff
@@ -684,14 +731,48 @@ func parseIndividualArg(_ arg: String, at position: Int) throws(ParserError)
...
+  ///     - arguments: The input from the command line.
+  ///     - responseFilePrefix: A leading character representing a response
+  ///       file argument
   ///
   /// - Throws: If parsing fails.
-  init(arguments: [String]) throws {
-    self.init(originalInput: arguments)
+  init(
+    arguments: [String],
+    responseFilePrefix: Character = "@"
```
Comment (rauhul) [inline Sources/ArgumentParser/Parsing/SplitArguments.swift]: please no default set params (unless public api)

**Comment by bkhouri** [comment]
this is a rather large PR. I will look to split it so the "source location" is a parent change to this.

**Comment by bkhouri** [comment]
https://github.com/apple/swift-argument-parser/pull/931 augments the source location information.

File: Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md (hunk 4 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Documentation.docc/Articles/ResponseFiles.md:336]: Thanks.

File: Sources/ArgumentParser/Parsing/SplitArguments.swift (hunk 8 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/SplitArguments.swift]: ACK.. I'll update this to make the response file prefix configurable, with `"@"` being the default.

**Comment by allevato** [comment]
Apologies if I missed it, but can you add a test for quoted empty string behavior in response files?

For example, these files:

```
-foo
''
-bar
```
and
```
-foo
""
-bar
```
should both expand to three arguments: `["-foo", "", "-bar"]`.

This was on my mind because a few months ago I fixed a bug in LLVM's response file parser that was incorrectly dropping empty arguments (instead of preserving them as `gcc` would): https://github.com/llvm/llvm-project/pull/187566

In fact, some of the other test cases I added might also serve as good tests of quotation handling for this implementation, too.

**Comment by bkhouri** [comment]
> Apologies if I missed it, but can you add a test for quoted empty string behavior in response files?

Thanks for the feedback @allevato .  I added more test coverage around empty quoted string s, and some with a new line character (`\n`) in a quoted string.

File: Sources/ArgumentParser/Parsable Types/ParsableCommand.swift [hunk 9]
```diff
@@ -22,6 +22,30 @@ public protocol ParsableCommand: ParsableArguments {
...
+  ///
+  ///     struct MyTool: ParsableCommand {
+  ///       static var responseFilePrefix: Character? { "+" }
+  ///       // ...
+  ///     }
+  ///
+  /// With the above configuration, `mytool +args.txt` expands `args.txt`
+  /// as a response file, while `mytool @foo` is treated as a literal
+  /// argument.
+  static var responseFilePrefix: Character? { get }
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsable Types/ParsableCommand.swift]: Does this need to be in the `ParsableCommand` protocol or could it be moved to `CommandConfiguration`? That seems like its more natural home.

File: Sources/ArgumentParser/Parsing/CommandParser.swift [hunk 10]
```diff
@@ -12,6 +12,11 @@
 struct CommandError: Error {
   var commandStack: [ParsableCommand.Type]
   var parserError: ParserError
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/CommandParser.swift:21]: Could you document these as well?

File: Sources/ArgumentParser/Parsing/CommandParser.swift [hunk 11]
```diff
@@ -87,7 +96,7 @@ extension CommandParser {
   ) -> Tree<ParsableCommand.Type>? {
     guard let (origin, element) = split.peekNext(),
       element.isValue,
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/CommandParser.swift]:
```suggestion
```
The changed line below makes this one unnecessary…

File: Sources/ArgumentParser/Parsing/InputOrigin.swift [hunk 12]
```diff
@@ -113,15 +122,131 @@ extension InputOrigin {
...
+  }
+}
+
 extension InputOrigin.Element {
   static func < (lhs: Self, rhs: Self) -> Bool {
+    // Ordering (least → greatest):
+    //   .argumentIndex(a) < .argumentIndex(b) when a < b
+    //   .argumentIndex(_) < .responseFile(_,_) < .defaultValue
+    // Two `.responseFile` values are ordered by lexicographically
+    // comparing their flattened chain step arrays.
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/InputOrigin.swift:194]: IIUC, the arguments in a response file are expanded "in place", such that they take the position of the `@file` command-line argument that includes them. Are there any issues with that in-place expansion leading to an actual ordering of arguments that's different from this ordering, where all the arguments from expanded response files are positioned after _all_ command-line arguments?

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 13]
```diff
@@ -0,0 +1,578 @@
...
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2026 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+internal import Foundation
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift:15]: Let's use this standard check:

```suggestion
#if canImport(FoundationEssentials)
internal import FoundationEssentials
#else
internal import Foundation
#endif
```

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 14]
```diff
@@ -0,0 +1,578 @@
...
+    var hasResponseFile = false
+
+    for (argvIndex, argument) in arguments.enumerated() {
+      if isResponseFileArgument(argument) {
+        hasResponseFile = true
+        guard let fileName = extractResponseFileName(argument) else {
+          result.append(
+            .init(value: argument, chain: [.argv(index: argvIndex)]))
+          continue
+        }
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: Seems like this is double checking – let's move the `extract...` call up to be the initial condition of the `if`, and eliminate this unreachable guard condition.

```suggestion
      if let fileName = extractResponseFileName(argument) {
        hasResponseFile = true
```

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 15]
```diff
@@ -0,0 +1,578 @@
...
+    }
+
+    return (result, hasResponseFile)
+  }
+
+  /// Determines if an argument is a response file reference.
+  /// - Parameter argument: The argument to check
+  /// - Returns: True if the argument starts with the response file prefix
+  func isResponseFileArgument(_ argument: String) -> Bool {
+    guard argument.count > prefix.utf8.count else { return false }
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: This check is implicitly covered by the next one.

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 16]
```diff
@@ -0,0 +1,578 @@
...
+    let fileName = String(argument.dropFirst(prefix.utf8.count))
+    return !fileName.isEmpty
+  }
+
+  /// Extracts the filename from a response file argument.
+  /// - Parameter argument: The response file argument (e.g., "@file.txt")
+  /// - Returns: The filename portion, or nil if not a valid response file argument
+  func extractResponseFileName(_ argument: String) -> String? {
+    guard isResponseFileArgument(argument) else { return nil }
+    return String(argument.dropFirst(prefix.utf8.count))
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: This will be incorrect if `prefix` is non-ASCII.
```suggestion
    return String(argument.dropFirst(prefix.count))
```

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 17]
```diff
@@ -0,0 +1,578 @@
...
+      throw ResponseFileError.fileNotFound(fileURL)
+    }
+
+    // Read file content
+    let content: String
+    do {
+      content = try String(contentsOf: fileURL, encoding: .utf8)
+    } catch {
+      throw ResponseFileError.readError(fileURL, error)
+    }
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]:
```suggestion
    let content: String
    do {
      content = try String(contentsOf: fileURL, encoding: .utf8)
    } catch CocoaError.fileReadNoSuchFile {
      throw ResponseFileError.fileNotFound(fileURL)
    } catch {
      throw ResponseFileError.readError(fileURL, error)
    }
```

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 18]
```diff
@@ -0,0 +1,578 @@
...
+  /// Resolves a file name to an absolute `URL`, handling relative paths.
+  ///
+  /// - Parameters:
+  ///   - fileName: The file name to resolve
+  ///   - parentURL: Optional parent file `URL` for relative resolution
+  /// - Returns: The resolved absolute URL
+  fileprivate func resolveFileURL(
+    _ fileName: String,
+    relativeTo parentURL: URL? = nil
+  ) -> URL {
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift:313]: I think this method is duplicating an existing API: https://developer.apple.com/documentation/foundation/url/init(filepath:directoryhint:relativeto:)

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift [hunk 19]
```diff
@@ -0,0 +1,578 @@
...
+    return URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
+      .appendingPathComponent(fileName)
+  }
+
+  /// Returns `true` if `path` is an absolute filesystem path.
+  ///
+  /// On POSIX, that means it starts with `/`. On Windows, that includes
+  /// drive-rooted paths (`C:\...` / `C:/...`), UNC paths (`\\server\...`),
+  /// and drive-relative-absolute paths (starting with `\` or `/`).
+  fileprivate static func isAbsolutePath(_ path: String) -> Bool {
```
Comment (natecook1000) [inline Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: Let's try hard not to rewrite our own version of this logic – is this necessary if `resolveFileURL` is simplified?

File: Sources/ArgumentParserTestHelpers/TemporaryFileFixture.swift [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+// This source file is part of the Swift Argument Parser open source project
+//
+// Copyright (c) 2026 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
```
Comment (natecook1000) [inline Sources/ArgumentParserTestHelpers/TemporaryFileFixture.swift:14]:
```suggestion
#if canImport(FoundationEssentials)
internal import FoundationEssentials
#else
internal import Foundation
#endif
```

**Review by natecook1000 — COMMENTED** [review]
Thanks, @bkhouri! Initial feedback here – I haven't reviewed the tests yet, will do that next.

File: Sources/ArgumentParserTestHelpers/TemporaryFileFixture.swift (hunk 20 above)
Comment (bkhouri) [reply Sources/ArgumentParserTestHelpers/TemporaryFileFixture.swift:14]: ack

File: Sources/ArgumentParser/Parsable Types/ParsableCommand.swift (hunk 9 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsable Types/ParsableCommand.swift]: It does not.  The `CommandConfiguration` is indeed a more natural home.

File: Sources/ArgumentParser/Parsing/CommandParser.swift (hunk 10 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/CommandParser.swift:21]: Will do.

File: Sources/ArgumentParser/Parsing/CommandParser.swift (hunk 11 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/CommandParser.swift]: Right, not sure how I missed that 🤦🏽

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 13 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift:15]: ack

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 14 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: That's cleaner. thanks.

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 16 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: `prefix` is of type `Character`. I don't want to hardcode `1` here in the [unlikely] event, that the response files have a multi-character prefix.  So trying to future proof this by taking the `.count`.

I'll change this to 
```suggestion
    return String(argument.dropFirst(String(prefix).count))
```

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 15 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: ack/

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 17 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: ack.

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 19 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift]: ack.

File: Sources/ArgumentParser/Parsing/InputOrigin.swift (hunk 12 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/InputOrigin.swift:194]: I don't believe there are ordering issues.  Every expanded argument (CLI or response-file origin) is assigned a sequential post-expansion `InputIndex` in `SplitArguments.swift`, and all origins use `.argumentIndex(SplitArguments.Index)`. Natural `<` ordering on `.argumentIndex` therefore reflects the true in-place expansion order — a response-file arg at post-expansion index 5 sorts between CLI arg at index 4 and CLI arg at index 6. Chain provenance for user-facing error messages lives separately in `FormattingContext.responseFileChains` keyed by the same post-expansion InputIndex.

There are also tests that help verify the ordering:
 - in `ResponseFileExpanderTests.swift:112-238` that validate the ordering within the response files processing is as expected.
 - in `ResponseFileEndToEndTests.swift`, `responseFileWithMixedArgsLastWinsCLI:122` parses a sample CLI `[@file, --count 100]`, where the response file sets `--count 2`. The test asserts `count == 100`, which comes after the response file on the command line.
 - in `ResponseFileEndToEndTests.swift`, `responseFileWithMixedArgsLastWinsResponseFile:141` is the same as the previous, but reversed the order on the command line.


I added additional automated test that will validate the CLI argument expansion is as expected.

File: Sources/ArgumentParser/Parsing/ResponseFileExpander.swift (hunk 18 above)
Comment (bkhouri) [reply Sources/ArgumentParser/Parsing/ResponseFileExpander.swift:313]: I can't recall why I didn't use the existing API.

File: Tests/ArgumentParserEndToEndTests/ResponseFileEndToEndTests.swift (hunk 1 above)
Comment (bkhouri) [reply Tests/ArgumentParserEndToEndTests/ResponseFileEndToEndTests.swift]: added
